[aoti-fx] Initial AOTInductor FX

angelayi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #160766
* __->__ #160765

Using the existing WrapperFxCodegen backend, this PR prototypes an AOT version of it which will directly return a graph module.

How to use:
```python
exported_gm = torch.export.export(model, inp, dynamic_shapes=dynamic_shapes).module()
compiled_gm = torch._inductor.aot_compile(
    exported_gm, inp, options={"fx_wrapper": True, "compile_threads": 1}
)
assert torch.allclose(model(*inp), compiled_gm(*inp))
```

Example graph:
```
class GraphModule(torch.nn.Module):
    def forward(self, arg2_1: "f32[3, 3]"):
        # No stacktrace found for following nodes
        linear_weight: "f32[3, 3]" = self.linear_weight
        buf0: "f32[3]" = torch.empty_strided([3], [1], dtype = torch.float32, device = device(type='cuda', index=0))
        triton_kernel_wrapper_mutation = torch.ops.higher_order.triton_kernel_wrapper_mutation(kernel_idx = 0, constant_args_idx = 0, grid = [(1, 1, 1)], tma_descriptor_metadata = {}, kwargs = {'out_ptr0': buf0, 'xnumel': 3, 'XBLOCK': 4});  triton_kernel_wrapper_mutation = None
        buf1: "f32[3, 3]" = torch.empty_strided([3, 3], [3, 1], dtype = torch.float32, device = device(type='cuda', index=0))
        linear_weight_view: "f32[3, 3]" = torch.as_strided(linear_weight, [3, 3], [1, 3], 0);  linear_weight = None
        addmm: "f32[3, 3]" = torch.addmm(buf0, arg2_1, linear_weight_view, alpha = 1, beta = 1, out = buf1);  buf0 = arg2_1 = linear_weight_view = addmm = None
        return buf1
```

The motivation behind this is that backends like ExecuTorch/MTIA would like to use inductor's optimization technologies, but might have their own graph lowering pipelines so they might not want to use AOTI (which generates an so).

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

angelayi committed:
[aoti-fx] Initial AOTInductor FX

[ghstack-poisoned]

blaine-rister commented:
Awesome! This could be very useful for unifying MTIA's inference UX with GPUs and CPUs.

blaine-rister commented on torch/_inductor/compile_fx.py:
```diff
@@ -1454,7 +1461,10 @@ def codegen_and_compile(
                         "GraphLowering.compile_to_fn", log_pt2_compile_event=True
                     ):
-                        if graph.aot_mode:
+                        if graph.aot_mode and graph.fx_wrapper:
```
Are `graph.cpp_wrapper` and `graph.fx_wrapper` mutually exclusive? Longer term, it might make sense to refactor them into the some sort of ABC.

```suggestion
                        if graph.aot_mode and graph.fx_wrapper:
                            assert not graph.cpp_wrapper
```

angelayi replied:
Yup good point! I've refactored it so that fx_wrapper and cpp_wrapper are now exclusive.

angelayi committed:
Update on "[aoti-fx] Initial AOTInductor FX"


To use:
```python
ep = torch.export.export(model, inp, dynamic_shapes=dynamic_shapes)
gm = torch._inductor.aot_compile(
    ep.module(), inp, options={"fx_wrapper": True, "compile_threads": 1}
)
assert torch.allclose(model(*inp), gm(*inp))
```


[ghstack-poisoned]

angelayi committed:
Update on "[aoti-fx] Initial AOTInductor FX"


To use:
```python
ep = torch.export.export(model, inp, dynamic_shapes=dynamic_shapes)
gm = torch._inductor.aot_compile(
    ep.module(), inp, options={"fx_wrapper": True, "compile_threads": 1}
)
assert torch.allclose(model(*inp), gm(*inp))
```


[ghstack-poisoned]

angelayi committed:
Update on "[aoti-fx] Initial AOTInductor FX"


Using the existing WrapperFxCodegen backend, this PR prototypes an AOT version of it which will directly return a graph module.

How to use:
```python
exported_gm = torch.export.export(model, inp, dynamic_shapes=dynamic_shapes).module()
compiled_gm = torch._inductor.aot_compile(
    exported_gm, inp, options={"fx_wrapper": True, "compile_threads": 1}
)
assert torch.allclose(model(*inp), compiled_gm(*inp))
```

The motivation behind this is that backends like ExecuTorch/MTIA would like to use inductor's optimization technologies, but might have their own graph lowering pipelines so they might not want to use AOTI (which generates an so).

[ghstack-poisoned]
